include/envoy/registry/registry.h

```diff
@@ -132,24 +133,45 @@ template <class Base> class FactoryRegistry {
     return *factories;
   }
 
-  static void registerFactory(Base& factory, absl::string_view name) {
+  static absl::flat_hash_map<std::string, std::string>& deprecatedFactoryNames() {
+    static auto* deprecated_factory_names = new absl::flat_hash_map<std::string, std::string>;
+    return *deprecated_factory_names;
+  }
+
+  /**
+   * instead_value are used when passed name was deprecated.
+   */
+  static void registerFactory(Base& factory, absl::string_view name,
+                              absl::string_view instead_value = "") {
     auto result = factories().emplace(std::make_pair(name, &factory));
     if (!result.second) {
       throw EnvoyException(fmt::format("Double registration for name: '{}'", factory.name()));
     }
+    if (!instead_value.empty()) {
+      deprecatedFactoryNames().emplace(std::make_pair(name, instead_value));
+    }
   }
 
   /**
    * Gets a factory by name. If the name isn't found in the registry, returns nullptr.
    */
   static Base* getFactory(absl::string_view name) {
+    checkDeprecated(name);
     auto it = factories().find(name);
     if (it == factories().end()) {
       return nullptr;
     }
     return it->second;
   }
 
+  static void checkDeprecated(absl::string_view name) {
+    auto it = deprecatedFactoryNames().find(name);
+    const bool status = it != deprecatedFactoryNames().end();
+    if (status) {
+      ENVOY_LOG(warn, "{} is deprecated, use {} instead.", it->first, it->second);
```

Comment:
@Shikugawa @lizan @htuch shouldn't we also increment the deprecated stat somehow? I feel like we really have to do this as part of our policy and should. @alyssawilk WDYT?

Reply:
Reopened #8540, we should increment deprecated stat / disable that in `ENVOY_DISABLE_DEPRECATED_FEATURES` cases etc.